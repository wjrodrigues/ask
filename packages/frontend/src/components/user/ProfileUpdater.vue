<template>
  <v-card class="mx-auto" width="344" :title="$t('profile.title')">
    <v-container>
      <v-form v-model="form" @submit.prevent="">
        <v-row justify="center">
          <input
            class="d-none"
            type="file"
            name="photo"
            ref="photo"
            accept="image/png, image/gif, image/jpeg, image/ico, image/jpg"
            @change="preloadPhoto"
          />

          <v-col cols="4">
            <v-avatar color="info" size="64">
              <img
                ref="previewPhoto"
                src=""
                :class="{ 'd-none': !photo }"
                class="preview-photo"
              />
              <v-icon icon="fas fa-user" :class="{ 'd-none': photo }"></v-icon>
            </v-avatar>

            <br />

            <v-btn variant="plain" @click="selectPhoto" :disabled="loading">
              {{ $t("profile.form.photo") }}
            </v-btn>
          </v-col>
        </v-row>

        <v-text-field
          v-model="first_name"
          :readonly="loading"
          :rules="[required]"
          clearable
          :label="$t('profile.first_name')"
          type="text"
          variant="underlined"
          :error-messages="errors.first_name"
          name="first_name"
        ></v-text-field>

        <v-text-field
          v-model="last_name"
          :readonly="loading"
          clearable
          :label="$t('profile.last_name')"
          type="text"
          variant="underlined"
          :error-messages="errors.last_name"
          name="last_name"
        ></v-text-field>

        <v-card-actions>
          <v-row justify="end">
            <v-btn
              type="submit"
              :disabled="loading"
              :loading="loading"
              color="success"
            >
              {{ $t("form.save") }}
            </v-btn>
          </v-row>
        </v-card-actions>
      </v-form>
    </v-container>
  </v-card>
</template>

<script lang="ts">
export default {
  data: () => ({
    form: false,
    first_name: "",
    last_name: "",
    url_photo: "",
    photo: false,
    loading: false,
    errors: { first_name: "", last_name: "", photo: "" },
  }),
  methods: {
    required(value: String) {
      return !!value || this.$t("form.field_required");
    },
    selectPhoto() {
      const inputPhoto = this.$refs.photo as HTMLInputElement;
      inputPhoto.click();
    },
    preloadPhoto(event: Event) {
      const target = event.target as HTMLInputElement;
      const files = target.files;

      if (files && files.length > 0) {
        const targetElement = this.$refs.previewPhoto as HTMLImageElement;
        const file = files[0];

        if (this.validFile(file)) {
          this.preview(file, targetElement);
          this.photo = true;
        }
      }
    },
    preview(file: File, targetElement: HTMLImageElement) {
      let reader = new FileReader();

      reader.readAsDataURL(file);
      reader.onload = () => {
        targetElement.src = `${reader.result}`;
      };
    },
    validFile(file: File) {
      const extension = file.name.split(".").pop() || "";

      return ["png", "jpeg", "ico", "gif", "jpg"].includes(extension);
    },
  },
};
</script>

<style>
.preview-photo {
  width: 100%;
  height: 100%;
}
</style>
