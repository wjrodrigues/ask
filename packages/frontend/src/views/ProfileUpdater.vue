<template>
  <div class="h-screen">
    <div class="d-flex align-center justify-center h-75">
      <ProfileUpdater />
    </div>
  </div>
</template>

<script lang="ts">
import ProfileUpdater from "@/components/user/ProfileUpdater.vue";

export default {
  components: {
    ProfileUpdater,
  },
};
</script>
