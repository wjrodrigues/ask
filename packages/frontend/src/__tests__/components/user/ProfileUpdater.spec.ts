import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";

import ProfileUpdater from "@/components/user/ProfileUpdater.vue";
import { basic_mount, faker, base_urls, nock } from "@/__tests__/help";

describe("when render", () => {
  it("render form with all components", () => {
    const wrapper = mount(
      ProfileUpdater,
      basic_mount({ props: {}, plugins: [] })
    );

    expect(wrapper.element).toMatchSnapshot();
  });

  it("validate required input", async () => {
    const wrapper = mount(
      ProfileUpdater,
      basic_mount({ props: {}, plugins: [] })
    );

    const inputs = await wrapper.findAll("input[type='text']");
    await inputs.forEach(async (e) => e.setValue("1"));
    await inputs.forEach(async (e) => e.setValue(""));
    await wrapper.trigger("change");

    const error_messages = await wrapper.findAll(".v-messages > div");
    expect(error_messages.length).toEqual(1);
    error_messages.map((msg) =>
      expect(msg.text()).toEqual("Campo é obrigatório")
    );
  });

  it("preview image supported", async () => {
    const wrapper = mount(
      ProfileUpdater,
      basic_mount({ props: {}, plugins: [] })
    );

    await wrapper.find("button").trigger("click");
    const inputFile = await wrapper.find("input[name='photo']");

    expect(inputFile.attributes("type")).toEqual("file");
    const inputElement = inputFile.element as HTMLInputElement;

    inputElement.files = [
      new File([new ArrayBuffer(1)], "file.jpg", {
        type: "image/jpg",
      }),
    ] as unknown as FileList;

    await inputFile.trigger("change");

    const preview = await wrapper.find("img");
    expect(preview.element.classList.toString()).toEqual("preview-photo");
  });

  it("do not preview image", async () => {
    const wrapper = mount(
      ProfileUpdater,
      basic_mount({ props: {}, plugins: [] })
    );

    const inputFile = await wrapper.find("input[name='photo']");
    const inputElement = inputFile.element as HTMLInputElement;

    inputElement.files = [
      new File([new ArrayBuffer(1)], "file.tiff", {
        type: "image/tiff",
      }),
    ] as unknown as FileList;

    await inputFile.trigger("change");

    const preview = await wrapper.find("img");
    expect(preview.element.classList.toString()).toEqual(
      "d-none preview-photo"
    );
  });
});

describe("when submit form", () => {
  it("validates form errors", async () => {
    const wrapper = mount(
      ProfileUpdater,
      basic_mount({ props: {}, plugins: [] })
    );

    await wrapper.find("[name='first_name']").setValue(faker.name.firstName());
    await wrapper.find("[name='first_name']").setValue("");
    await wrapper.trigger("change");

    const error_messages = await wrapper.findAll(".v-messages > div");
    expect(error_messages[0].text()).toEqual("Campo é obrigatório");

    expect(wrapper.vm.errors).toEqual({
      first_name: "",
      last_name: "",
      photo: "",
    });
    expect(wrapper.vm.form).toEqual(false);
  });

  it("sends only first name", async () => {
    nock(base_urls.API_REGISTRATION).post("/users/profile").reply(200);

    const wrapper = mount(
      ProfileUpdater,
      basic_mount({ props: {}, plugins: [] })
    );

    await wrapper.find("[name='first_name']").setValue(faker.name.firstName());
    await wrapper.trigger("change");
    await wrapper.find("form").trigger("submit");

    const error_messages = await wrapper.findAll(".v-messages > div");
    expect(error_messages).toHaveLength(0);
  });

  it("sends with photo", async () => {
    nock(base_urls.API_REGISTRATION).post("/users/profile").reply(200);
    nock(base_urls.API_REGISTRATION)
      .post("/users/profile/presigned_url")
      .reply(
        200,
        "http://localhost/profilepictures/57825751-5412-4327-87de-dbcec6a52ef7.jpg?X-Amz"
      );
    nock("http://localhost")
      .put("/profilepictures/57825751-5412-4327-87de-dbcec6a52ef7.jpg?X-Amz")
      .reply(200);

    const wrapper = mount(
      ProfileUpdater,
      basic_mount({ props: {}, plugins: [] })
    );

    await wrapper.find("[name='first_name']").setValue(faker.name.firstName());

    const inputFile = await wrapper.find("input[name='photo']");
    const inputElement = inputFile.element as HTMLInputElement;

    inputElement.files = [
      new File([new ArrayBuffer(1)], "file.jpg", { type: "image/jpg" }),
    ] as unknown as FileList;

    await wrapper.find("form").trigger("submit");

    const error_messages = await wrapper.findAll(".v-messages > div");
    expect(error_messages).toHaveLength(0);
  });

  it("updates profile error", async () => {
    nock(base_urls.API_REGISTRATION).post("/users/profile").reply(500);

    const wrapper = mount(
      ProfileUpdater,
      basic_mount({ props: {}, plugins: [] })
    );

    await wrapper.find("[name='first_name']").setValue(faker.name.firstName());
    await wrapper.trigger("change");
    await wrapper.find("form").trigger("submit");
  });
});
