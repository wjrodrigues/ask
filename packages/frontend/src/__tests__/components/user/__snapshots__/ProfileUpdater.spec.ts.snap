// Vitest Snapshot v1

exports[`when render > render form with all components 1`] = `
<div
  class="v-card v-theme--light v-card--density-default v-card--variant-elevated mx-auto"
  style="width: 344px;"
>
  <!---->
  <div
    class="v-card__loader"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      class="v-progress-linear v-theme--light"
      role="progressbar"
      style="height: 0; --v-progress-linear-height: 2px;"
    >
      <!---->
      <div
        class="v-progress-linear__background"
        style="width: 100%;"
      />
      <transition-stub
        appear="false"
        css="true"
        name="fade-transition"
        persisted="false"
      >
        <div
          class="v-progress-linear__indeterminate"
        >
          
          <div
            class="v-progress-linear__indeterminate long"
          />
          <div
            class="v-progress-linear__indeterminate short"
          />
          
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <div
    class="v-card-item"
  >
    <!---->
    <div
      class="v-card-item__content"
    >
      <div
        class="v-card-title"
      >
        Perfil
      </div>
      <!---->
      <!---->
    </div>
    <!---->
  </div>
  <!---->
  
  <div
    class="v-container"
  >
    <form
      class="v-form"
      novalidate=""
    >
      
      <div
        class="v-row justify-center"
      >
        <input
          accept="image/png, image/gif, image/jpeg, image/ico, image/jpg"
          class="d-none"
          name="photo"
          type="file"
        />
        <div
          class="v-col v-col-4"
        >
          <div
            class="v-avatar bg-info v-avatar--density-default v-avatar--variant-flat"
            style="width: 64px; height: 64px;"
          >
            
            <img
              class="d-none preview-photo"
              src=""
            />
            <i
              aria-hidden="true"
              class="fas fa-user mdi v-icon notranslate v-theme--light v-icon--size-default"
            />
            
            
            <!---->
            <span
              class="v-avatar__underlay"
            />
            
          </div>
          <br />
          <button
            class="v-btn v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-plain"
            type="button"
          >
            
            <span
              class="v-btn__overlay"
            />
            <span
              class="v-btn__underlay"
            />
            
            <!---->
            <span
              class="v-btn__content"
              data-no-activator=""
            >
              
              
              Selecionar imagem
              
              
            </span>
            <!---->
            <!---->
          </button>
        </div>
      </div>
      <div
        class="v-input v-input--horizontal v-input--density-default v-text-field v-text-field--flush-details"
      >
        <!---->
        <div
          class="v-input__control"
        >
          
          <div
            class="v-field v-field--appended v-field--variant-underlined v-theme--light"
            role="textbox"
          >
            <div
              class="v-field__overlay"
            />
            <div
              class="v-field__loader"
            >
              <div
                aria-valuemax="100"
                aria-valuemin="0"
                class="v-progress-linear v-theme--light"
                role="progressbar"
                style="height: 0; --v-progress-linear-height: 2px;"
              >
                <!---->
                <div
                  class="v-progress-linear__background"
                  style="width: 100%;"
                />
                <transition-stub
                  appear="false"
                  css="true"
                  name="fade-transition"
                  persisted="false"
                >
                  <div
                    class="v-progress-linear__indeterminate"
                  >
                    
                    <div
                      class="v-progress-linear__indeterminate long"
                    />
                    <div
                      class="v-progress-linear__indeterminate short"
                    />
                    
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <div
              class="v-field__field"
              data-no-activator=""
            >
              <!---->
              <label
                class="v-label v-field-label"
                for="input-1"
              >
                <!---->
                
                Nome
                
              </label>
              
              
              <!---->
              <input
                class="v-field__input"
                id="input-1"
                name="first_name"
                size="1"
                type="text"
              />
              <!---->
              
              
            </div>
            <transition-stub
              appear="false"
              css="true"
              name="expand-x-transition"
              persisted="false"
            >
              <div
                class="v-field__clearable"
                style="display: none;"
              >
                <i
                  aria-hidden="false"
                  aria-label="Clear Nome"
                  class="mdi-close-circle mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable"
                  role="button"
                />
              </div>
            </transition-stub>
            <div
              class="v-field__append-inner"
            >
              <!---->
              <!---->
            </div>
            <div
              class="v-field__outline"
            >
              <!---->
              <label
                aria-hidden="true"
                class="v-label v-field-label v-field-label--floating"
                for="input-1"
              >
                <!---->
                
                Nome
                
              </label>
            </div>
          </div>
          
        </div>
        <!---->
        <div
          class="v-input__details"
        >
          <transition-group-stub
            appear="false"
            class="v-messages"
            css="true"
            name="slide-y-transition"
            persisted="false"
            tag="div"
          >
            <!---->
          </transition-group-stub>
          <!---->
        </div>
      </div>
      <div
        class="v-input v-input--horizontal v-input--density-default v-text-field v-text-field--flush-details"
      >
        <!---->
        <div
          class="v-input__control"
        >
          
          <div
            class="v-field v-field--appended v-field--variant-underlined v-theme--light"
            role="textbox"
          >
            <div
              class="v-field__overlay"
            />
            <div
              class="v-field__loader"
            >
              <div
                aria-valuemax="100"
                aria-valuemin="0"
                class="v-progress-linear v-theme--light"
                role="progressbar"
                style="height: 0; --v-progress-linear-height: 2px;"
              >
                <!---->
                <div
                  class="v-progress-linear__background"
                  style="width: 100%;"
                />
                <transition-stub
                  appear="false"
                  css="true"
                  name="fade-transition"
                  persisted="false"
                >
                  <div
                    class="v-progress-linear__indeterminate"
                  >
                    
                    <div
                      class="v-progress-linear__indeterminate long"
                    />
                    <div
                      class="v-progress-linear__indeterminate short"
                    />
                    
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <div
              class="v-field__field"
              data-no-activator=""
            >
              <!---->
              <label
                class="v-label v-field-label"
                for="input-3"
              >
                <!---->
                
                Sobrenome
                
              </label>
              
              
              <!---->
              <input
                class="v-field__input"
                id="input-3"
                name="last_name"
                size="1"
                type="text"
              />
              <!---->
              
              
            </div>
            <transition-stub
              appear="false"
              css="true"
              name="expand-x-transition"
              persisted="false"
            >
              <div
                class="v-field__clearable"
                style="display: none;"
              >
                <i
                  aria-hidden="false"
                  aria-label="Clear Sobrenome"
                  class="mdi-close-circle mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable"
                  role="button"
                />
              </div>
            </transition-stub>
            <div
              class="v-field__append-inner"
            >
              <!---->
              <!---->
            </div>
            <div
              class="v-field__outline"
            >
              <!---->
              <label
                aria-hidden="true"
                class="v-label v-field-label v-field-label--floating"
                for="input-3"
              >
                <!---->
                
                Sobrenome
                
              </label>
            </div>
          </div>
          
        </div>
        <!---->
        <div
          class="v-input__details"
        >
          <transition-group-stub
            appear="false"
            class="v-messages"
            css="true"
            name="slide-y-transition"
            persisted="false"
            tag="div"
          >
            <!---->
          </transition-group-stub>
          <!---->
        </div>
      </div>
      <div
        class="v-card-actions"
      >
        
        <div
          class="v-row justify-end"
        >
          <button
            class="v-btn v-theme--light text-success v-btn--density-default v-btn--size-default v-btn--variant-text"
            type="submit"
          >
            
            <span
              class="v-btn__overlay"
            />
            <span
              class="v-btn__underlay"
            />
            
            <!---->
            <span
              class="v-btn__content"
              data-no-activator=""
            >
              
              
              Salvar
              
              
            </span>
            <!---->
            <!---->
          </button>
        </div>
        
      </div>
      
    </form>
  </div>
  
  <!---->
  
  <!---->
  <span
    class="v-card__underlay"
  />
  
</div>
`;
